### DISCUSSION TITLE
Why are concrete values used to check inequality?

### DISCUSSION THREAD

**Discussion by andrew222651** [pschanely/CrossHair#100] [Q&A] [answered]
Suppose I check the following function
```python
def f(x: int) -> int:
    '''
    post: x < __return__
    '''
    return x + 1
```
Then in the first iteration, the debug log shows that while executing the function body, CrossHair calls `choose_possible(0 == x_2)` and later we see `SMT realized symbolic: x_2 + 1 == 1`. In later iterations, `x_2` is set to other values, incrementing each time. Why isn't it kept as a symbolic integer? Then only one iteration should be needed.

**Comment by pschanely** [comment, marked as answer]
Hi! Agree that only one iteration should be needed. Is this the exact code you're using? When [I run it](https://crosshair-web.org/?crosshair=0.1&python=3.8&flags=verbose&gist=79dfeb8773b474a977289ac775f2ff6c), I only get one iteration:
```
|  |check() Check  main
|      |analyze_function() Analyzing  f
|       |condition_parser() Using parsers:  (<AnalysisKind.PEP316: 'PEP316'>, <AnalysisKind.icontract: 'icontract'>, <AnalysisKind.asserts: 'asserts'>)
|    |analyze() Analyzing postcondition: " x < __return__ "
|    |analyze() assuming preconditions:  
|     |condition_parser() Using parsers:  (<AnalysisKind.PEP316: 'PEP316'>, <AnalysisKind.icontract: 'icontract'>, <AnalysisKind.asserts: 'asserts'>)
|     |analyze_calltree() Begin analyze calltree  f
|     |analyze_calltree() Iteration  1
|       |gen_args() created proxy for x as type: SmtInt
|        |choose_possible() SMT chose: x_2 < x_2 + 1
|      |attempt_call() Postcondition confirmed.
|     |analyze_calltree() Path tree stats StateSpaceCounter({<VerificationStatus.CONFIRMED: 2>: 1})
|     |analyze_calltree() Iter complete. Worst status found so far: CONFIRMED
|     |analyze_calltree() Exhausted calltree search with CONFIRMED and 0 messages. Number of iterations:  1
```
Possibly related to a difference in CrossHair versions?
Another idea: if you are trying to, for example, print any intermediate values, you'll force realization of the integer which would cause the kind of behavior you describe.

**Reply by andrew222651** [reply]
Aha, I had added `print(lcls)` [here](https://github.com/pschanely/CrossHair/blob/master/crosshair/core.py#L1301) and removing that solves the problem. Is there a way to get a string representation of that stuff without realizing?

**Reply by pschanely** [reply]
Sadly, not at present. One tension here: we mostly want symbolics to be indistinguishable from their native counterparts. So, for example, we would not simply change `__str__` on `SymbolicInt` to return something about the symbolic state. After all, someone might be trying to verify something about their function that takes the string representation of an integer!

That said we could probably special-case something, somewhere ... possibly in CrossHair's own `debug()` utility. Feel free to drop a quick feature request!
